deriveY/deriveY.md

```diff
@@ -0,0 +1,58 @@
+# `deriveY` and quadratic residue
+
+Suppose we have an elliptic curve *E* over *Z<sub>P</sub>*, with *P* a prime, given by the equation *y<sup>2</sup>=x<sup>3</sup>+ax<sup>2</sup>+b*. Then knowing *x* there are two results from the equation of the curve, *y* and *-y*. The parity byte of *y*, 0x02 even, 0x03 odd, is enough to know which of the two results is the *y* we need. This is what the `deriveY` function does, revels *y* having as inputs the compressed form of *y* and *x*. 
+The main part of the function relies on computing the quadratic residue modulus *P*, this is resolving something like *y<sup>2</sup> = n mod (P)*. This can be done using different algorithms, the library `EllipticCurve.sol` uses the computation of *n<sup>(p + 1)/4</sup>*, but this only works for Elliptic Curves having  `P = 3 mod(4)`.
+
+```solidity
+pragma solidity ^0.5.0;
+
+import "./EllipticCurve.sol";
+
+contract Secp256k1 is EllipticCurve {
+
+    function deriveY(
+    uint8 prefix,
+    uint256 x,
+    uint256 a,
+    uint256 b,
+    uint256 pp)
+  public pure returns (uint256 y)
+  {
+    // x^3 + ax + b
+    uint256 y2 = addmod(mulmod(x, mulmod(x, x, pp), pp), addmod(mulmod(x, a, pp), b, pp), pp);
+    uint256 y_ = expMod(y2, (pp + 1) / 4, pp);
+    // uint256 cmp = yBit ^ y_ & 1;
+    y = (y_ + prefix) % 2 == 0 ? y_ : pp - y_;
+  }
+}
+```
+
+## Case `p=1(4)`
+
+For Elliptic Curves in which `P=1(4)` there are others algorithms that compute the modular square root. In fact they are more general algorithms so they could be implemented for all the Elliptic Curves the library supports, but the computational costs are higher than for the current function `deriveY`.
+
+The most commun algorithms for computing the *n* square root module *P*  are the *Tonelli-Shanks* and the *Cipolla*'s algorithm. Because in the `EllipticCurve.sol` library the only curve that satisfies `P= 1 mod 4` is `Secp224r1` we will focus on its characteristics. In this curve the value of *P* is 2<sup>224</sup> -2<sup>96</sup>+1, so *P = 1 (4)*.
```

Comment:
```suggestion
The most common algorithms for computing the *n* square root module *P*  are the *Tonelli-Shanks* and the *Cipolla*'s algorithm. Because in the `EllipticCurve.sol` library the only curve that satisfies `P= 1 mod 4` is `Secp224r1` we will focus on its characteristics. In this curve the value of *P* is 2<sup>224</sup> -2<sup>96</sup>+1, so *P = 1 mod (4)*.
```